### DISCUSSION TITLE
USB Power Trigger

### DISCUSSION THREAD

**Discussion by ryancardoza** [gopro/labs#1180] [Q&A]
Is there a way to detect if the SD card is not installed and to shut down the GoPro13 when power is turned on?

This is an aviation cockpit recording project. Attached is my setup.

![IMG_7036](https://github.com/user-attachments/assets/20c0ddf8-7d58-4fbb-8ca3-7e7dd41b3038)

My current settings:

**Set Time**
https://gopro.github.io/labs/control/precisiontime/

**Set USB Power Trigger** 
(Power on: camera boots, records. Power off: stops recording, powers off.)
https://gopro.github.io/labs/control/usb/
*FAST=1*WAKE=2*BOOT="!Lbt"!SAVEbt=<u0!X<r0!S>u0=At:B<u0>r0=Bt:B+=B-A>B10>r0!E+!1N+!1O<r0!X!R10

**Sets Owner, Aircraft, & Contact Information**
https://gopro.github.io/labs/control/owner/
*OWNR="ABC"

**Sets File Name**
https://gopro.github.io/labs/control/basename/
*BASE="ABCD-EFGHI-yymmdd-"

**Add an Overlay to Video (Permanent Overlay)**
https://gopro.github.io/labs/control/overlays/
g1!MLFIN=0!MBRNO=10!MBURN="(0,40)©ABC\n[TCHH:MM:SS\nmm-dd-yyyy ]"

**Log GPS**
*GPSL=1

**32GB Chapters**
*64BT=1

**Bypass common pop-ups**
*BYPS=1

**GPS time/timecode sync**
*SYNC=1

**Trust USB power**
*TUSB=1

**Disable QR code detection when recording**
*QRDR=0

Thank you,
Ryan

**Comment by dnewman-gpsw** [comment]
Testing for the sd card is a missing Labs feature and a good idea. I'm adding that now. 

Notes on your modes: 
- for 32GB chapters your need *64BT=32000  (*64BT=1 is for older cameras to get 12GB.)
- *QRDR=0 is the default, means do not QR code scan during capture. No need to set if it if this is the behavior you want.

**Reply by dnewman-gpsw** [reply]
There is a workaround for camera that is running a scripts that start recording automatically.

After !S <- start

You can immediately test if the camera is recording, if not, no sd card, then off.

In your script here: ```<r0!S```
replace with: ```<r0!S<r0!O```

**Reply by ryancardoza** [reply]
David,

Thank you for your feedback. I have made the 64BT and QRDR changes.

For the script modification, will the script run if the SD card is not inserted? My understanding is that the GoPro is dependent on the SD card because the script is saved to it instead of the GoPro. I modified that script, but it did not turn off the camera when the SD card was not installed.

Is there a way to store a script in the GoPro memory, where the permanent metadata lives?

**Format SD Card**
When the SD card extender cable is inserted into the GoPro, but there is not an SD card in the extender, the GoPro thinks there is an SD card installed (I think the SD card locking mechanism triggers the GoPro to think it has an SD card in it) and offers to format the card. I can see how this would be desirable in 99% of cases when there is a card in the GoPro and the camera cannot read it. But with the extender, the SD card can be removed from the extender, but the extender still installed in the GoPro, activating the lock trigger and therefore triggering the format SD card function.

Ryan

**Reply by dnewman-gpsw** [reply]
Yes you can save the script in the cameras internal flash, we just don't recommend doing that, as if you get the script wrong you could have a camera stuck. There are ways out, just not as easy as ejecting an SD card. As you script has an exit plan (when not powered by USB) it is pretty safe to do this.

```*FAST=1*WAKE=2*BOOT="<u0!X<r0!S<r0!O>u0=At:Br0=Bt:B+=B-A>B10>r0!E+!1N+!1O<r0!X!R15"```

**Reply by ryancardoza** [reply]
David, that is very helpful. Thank you for sharing your expertise!

There are scenarios where we will turn the power off in the aircraft to simulate emergencies. I would like these events to continue to be recorded, but do not want the camera to run off the battery unnecessarily long.

Would there be a way to modify the script so that the camera continues to record even **if the power is turned off** as long as the GoPro is in motion? When the GoPro detects it is no longer in motion for X seconds, then it would wait 10 seconds stop recording and shut down. If the GoPro power is restored, it continues with the script until the power is turned off.

**Reply by dnewman-gpsw** [reply]
Yes that is a doable. You, or a coding friend, will just need to design the conditions for what is motion, scene, gyro, accelerometer, GPS, are all options.  All can be created in the Blocky visual script builder https://gopro.github.io/labs/build/

**Reply by ryancardoza** [reply]
David, I was able to add this feature by resetting variable A if motion is detected.

`*FAST=1*WAKE=2*BOOT="<u0!X<r0!S>u0=At:B<u0>g5=At:B>r0=Bt:B+=B-A>B9>r0!E+!1N+!1O<r0!1N+!1O!R10"`

<img src="https://github.com/user-attachments/assets/8bcaa4d7-e152-44dc-8e58-41a8c08124b0" alt="Aviation Cockpit Video V1" width="200"/>

<img src="https://github.com/user-attachments/assets/4448f7b4-2e60-4e6d-84d1-5757efc88b1e" alt="Aviation Cockpit Video V1 QR" width="200"/>

Three questions:
1. What do you recommend if the QR code has too much information for the camera to be able to scan the code? Break it into multiple images?
2. If the GoPro doesn't have an SD card, or the SD card is corrupted, does the BOOT script run? I have set the camera to auto power off after 1-minute which conveniently shuts the camera down, but I would prefer that the camera shutdown if the memory card is not inserted or if it is corrupted (the result of the external SD card reader being empty but inserted into the camera). Ideally, I would script the camera to attempt to format (!FORMAT) the card without intervention, if that is unsuccessful, then to shut down.
3. With *WAKE=2, if the camera shuts down with the USB power on (say the SD card is corrupt), when the power is turned off the camera turns back on. Is that the expected outcome?

I appreciate your time.

Ryan

**Reply by dnewman-gpsw** [reply]
1. There seems to be a bug in Blocky, remove the comment (blue question mark) as that is cause the QR code to explode (and making an invalid command.)  There are ways to split up QR code in subroutines.

2. The test for SD card was a hack by using capture, the next Labs up with include a variable for storage space, or it will return -1 on a SD error or missing.

3. Yes.

**Comment by dnewman-gpsw** [comment]
Why are you running a micro SD card extender cable?

**Reply by ryancardoza** [reply]
In our use case, it is simpler to have our flight instructors swap the SD card in between flights than to: remove the camera, open the side door, remove the SD card, replace the SD card, and reinstall the camera for the next flight.

We will assign each of our instructors and SD card. ~The only caveat at the moment is that each card will need to have the "bt" script copied onto it otherwise it will not auto-start the recording or shutoff when the power is turned off.~

@dnewman-gpsw explained that the script can be saved directly to the internal camera metadata. Example:
`*FAST=1*WAKE=2*BOOT="<u0!X<r0!S>u0=At:B<u0>g5=At:B>r0=Bt:B+=B-A>B9>r0!E+!1N+!1O<r0!1N+!1O!R10"`

I'm not sure if GoPro has done durability testing on the SD card slot, USB-C port or the side door to determine how many cycles they can take before failure. Using the extender should theoretically increase the lifespan by decreasing the cycles on those items.
